### DISCUSSION TITLE
what are Remappings/foundry.toml and how to know when to use them?

### DISCUSSION THREAD

**Discussion by SquilliamX** [Cyfrin/foundry-full-course-cu#692] [Q&A] [answered]
Hi everyone, I see time and time again that we are adjusting our remapping in foundry.toml. If I was writing my own contract how would I know if I need to adjust my remapping in foundry.toml? Also what is foundry.toml? Also if I needed to add remapping to my foundry.toml, how would I know what to remap it to exactly? In fact when you do something like `` remapping = ['@openzeppelin/contracts=lib/openzeppelin-contracts/contracts']`` what is this line of code even doing?

**Comment by theirrationalone** [comment, marked as answer]
As we installed so many libraries/packages, Actually we dont't need to remap any path for our local development & testing protocol contracts inside our own Yard, why????, By installing them we can Go straight forward and can import directly into our desired places with absolute/relative(whatever you wish to) pathing technique. 

Since our main/created by us, protocol contracts also use / import these different libraries if they have their need accordingly for e.g.,

1. AggregatorV3Interface to get different PriceFeeds offchain data from chainlink oracle, 

2. ERC20 tokens to get dsc for dscEngine and again PriceFeeds too, 

3. ERC721 NFT Tokens to implement our own NFTs, 

4. VRFCoordinators, AutomationCompatible to get a time based winner of our lottery automatically, etc... 

You might noticed, we often require their mocks to work with them locally for testing purposes. Why??? Because we can't perform those transactions on mainnet, actually we can if we have enough money in our wallet. I'm assuming that we don't have enough money in our wallet although we can perform transactions onto testnet i.e., sepolia, goerli.


Likewise, let's say, We deployed our Raffle/Lottery contract protocol either on mainnet Or testnets and what mainnet Or testnets actually are?? They are real live chains. So as a reaction our absolute/relative pathing import technique won't work there.. Whyyy??? Since we're/our Raffle contract, is on a live chain, It required offchain data + logic and we have given  | implemented absolute/relative path import to retrieve required module|function|data|contract|library|interface but due to absolute/relative path import our contract will try to import required things either from its root Or from where it was placed, Therefore our contract will not work if we try absolute/relative import. 

You might be thinking we have our  `lib` then so why we didn't also deployed all libraries into??? It will cost tooo muchhh, we only need to deploy only those contracts which are required by our implementation i.e., VRFCoordinator, aggregatorV3Interface, etc. 

Now Blockchain has its own way to import things.... 
1. Each and every contract that we deployed will be allocated parallelly one after one. 
2. `@` might indicates that the contract we are importing is an external dependency and can make an api call if needed to fetch offchain data + logic, from the protocol that was originally deployed by chainlink, openzeppelin, solmate, etc.
3. We can say it's absolute importing technique but onto the live chain in a block. 
4. Might also helps to maintain nonce, transactions records, owner signing information. 


So that's why we need to have remappings to provide accurate & exact path on mainnet Or testnets.

**Reply by SquilliamX** [reply]
Local Development & Testing:

For contracts developed and tested locally, you don't necessarily need remappings. This is because all the required files and libraries are accessible on your local machine. You can directly use absolute or relative paths to import other contracts or libraries.
Utilizing External Libraries:

Contracts frequently depend on established, well-tested libraries, like those offered by OpenZeppelin, Chainlink, etc. Some examples include:
a. AggregatorV3Interface: Used to retrieve off-chain data from Chainlink oracles.
b. ERC20: A standard representation for fungible tokens.
c. ERC721: A standard for creating non-fungible tokens (NFTs).
d. VRFCoordinators, AutomationCompatible: Interfaces/contracts provided by Chainlink for generating random numbers and automating specific tasks.
Mocks for Local Testing:

Mock contracts are often used in local testing environments to emulate contract behaviors without interacting with their mainnet versions. This approach avoids the often high costs associated with mainnet interactions because executing transactions on the mainnet requires actual ETH.
Deployment on Mainnet/Testnets:

When you're ready to deploy your contracts on a live blockchain network (whether it's the mainnet or a testnet), the conventional file paths become irrelevant. This is the juncture where remappings become crucial.
The blockchain doesn't operate with traditional file paths. Instead, it functions based on addresses and deployed contract instances. When your contract requires interaction with another contract, it can't employ a local file path to do so.
The Expense of Deploying Libraries:

Deploying every library contract to the mainnet can be prohibitively expensive. Instead, developers typically deploy only those contracts or interfaces that are directly required for their project.
Blockchain's Method of Importing:

On-chain contracts don't "import" in the way we understand in traditional programming languages. They reference other contracts through their specific addresses.
In the realm of remappings, the @ symbol signifies an external dependency. While this doesn't imply an API call in the conventional sense, it does indicate a reference to an externally deployed contract.
The Role of Remappings:

Remappings are instrumental in providing an accurate reference to the right on-chain addresses or contract instances when deploying or managing contracts on the mainnet or testnets. They bridge the gap between the familiarity of local development and the peculiarities of on-chain execution, ensuring that contracts function as intended when deployed on a live blockchain.

**Comment by theirrationalone** [comment]
Beware of community guidelines. 😉

**Reply by SquilliamX** [reply]
Oh sorry I was still confused so copy and pasted what you said into AI so they can break it down even further and I didn't wanna ask dumb questions lol. I put what AI said in the thread just in case anyone else was reading this thread and needed help they could just read it. What community guidelines are you talking about?
